src/charm.py

```diff
@@ -0,0 +1,139 @@
+        self.framework.observe(self.on.build_langpacks_action, self._on_build_langpacks)
+        self.framework.observe(self.on.upload_langpacks_action, self._on_upload_langpacks)
+        self.framework.observe(self.on.stop, self._on_stop)
+
+        self._langpacks = Langpacks()
+
+    def _on_start(self, event: ops.StartEvent):
+        """Handle start event."""
+        self.unit.status = ops.MaintenanceStatus("Updating langpack-o-matic checkout")
+
+        try:
+            self._langpacks.update_checkout()
+        except CalledProcessError:
+            self.unit.status = ops.BlockedStatus(
+                "Failed to start services. Check `juju debug-log` for details."
+            )
+            return
+
+        self.unit.status = ops.ActiveStatus()
+
+    def _on_install(self, event: ops.InstallEvent):
+        """Handle install event."""
+        self.unit.status = ops.MaintenanceStatus("Installing langpack dependencies")
+        try:
+            self._langpacks.install()
+        except (CalledProcessError, PackageError, PackageNotFoundError):
+            self.unit.status = ops.BlockedStatus(
+                "Failed to install packages. Check `juju debug-log` for details."
+            )
+            return
+
+        self.unit.status = ops.MaintenanceStatus("Setting up crontab")
+        self._langpacks.setup_crontab()
+
+        self.unit.status = ops.ActiveStatus()
+
+    def _on_config_changed(self, event: ops.ConfigChangedEvent):
+        """Update configuration and fetch code updates."""
+        self.unit.status = ops.MaintenanceStatus("Importing signing key")
+
+        try:
+            secret_id = self.config["gpg-secret-id"]
+        except KeyError:
+            logger.warning("No 'gpg-secret-id' config, can't set up signing key")
+            self.unit.status = ops.ActiveStatus(
+                "No 'gpg-secret-id' config, signing will not be available."
+            )
+            return
+
+        try:
+            gpgkey: Secret = self.model.get_secret(id=secret_id)
+            keycontent = gpgkey.get_content().get("key")
+        except ops.SecretNotFoundError:
+            logger.warning("Signing key secret not found")
+            self.unit.status = ops.ActiveStatus(
+                "Secret not available. Check that access was granted."
+            )
+            return
+
+        try:
+            self._langpacks.import_gpg_key(keycontent)
+        except CalledProcessError:
+            self.unit.status = ops.BlockedStatus(
+                "Failed to import the signing key. Check `juju debug-log` for details."
+            )
+            return
+
+        self.unit.status = ops.ActiveStatus()
+
+    def _on_build_langpacks(self, event: ops.ActionEvent):
+        """Build new langpacks."""
+        release = event.params["release"]
+        base = event.params["base"]
+
+        self.unit.status = ops.MaintenanceStatus(
+            f"Building new langpacks, base={base}, release={release}"
```

Comment:
I'd maybe shorten this to "Building langpacks. See `juju debug-log`"

We want status messages to be short and sharp, giving a clear indication of required action.

Same applies elsewhere.